oo.py

```diff
@@ -0,0 +1,276 @@
+
+#this imports conduce.api
+import api as func_api
+
+from requests.packages.urllib3.exceptions import InsecureRequestWarning
+requests.packages.urllib3.disable_warnings(InsecureRequestWarning)
+
+
+DEV_SERVER = "https://dev-app.conduce.com"
+
+#Resources
+USER_ASSET = "userassets"
+TEMPLATE = "templates"
+
+
+
+_LIST_KEYS = {
+    USER_ASSET: "asset",
+    TEMPLATE: "template",
+}
+
+
+class ConduceError(Exception):
+    def __init__(self, url, resp, msg=None):
+        self.code = resp.status_code
+        self.url = url
+        self.text = resp.text
+        msg = "Got a error code %s from %s: %s" % (resp.status_code, url, resp.text)
+        super(ConduceError, self).__init__(msg)
+
+
+class ConduceAPIClient(object):
+    #Simple Version history:
+    # 1.0 Version including all features to migrate using sluice-to-banksy & run a RT daemon
+    # 2.0: Use v2 datasets for everything
+    # 2.1: Make create dataset sage function & ensure_sage_dataset function
+    # 3.0: Integrate with functional api
+    __version = 3.0
+
+    def __init__(self, host="https://app.conduce.com", api_key=None, email=None, password=None, insecure=True):
+        """
+        Configure an API client that can connect to Conduce servers.
+        You may provide either an API key or an email and password.  The API key
+        is strongly encouraged since it does not expose a user's credentials, and
+        does not provide full access to a user's account if compromised.
+        If email/password are provided, you must call the login() function to
+        get a cookie before making API calls.
+
+        :param host: optional host, only used if connecting to beta/preview servers
+        :param api_key: optional api key.  This is the recommended way to use this client
+        :param email: optional email if using login
+        :param password: optional pasword if using login
+        :rtype: ConduceAPIClient
+        """
+        #TODO: Integrate this with config.py
+        if api_key is None:
+            #using email & password
+            if email is None and password is None:
+                raise ValueError("api_key or email & password required")
+            if email is None or password is None:
+                raise ValueError("both email and password must be provided")
+        else:
+            #If an api key is specified, make sure email & password are none
+            if email is not None or password is not None:
+                #TODO: Use the warning lib
+                print "Warning, ignoring email and password inputs because api key provided"
+            email = None
+            password = None
+        #TODO: Make sure API has valid characters letters, numbers & - & _
+        self._api_key = api_key
+        self._email = email
+        self._password = password
+        self._cookies = None
+        self._verify = not insecure
+        info = urlparse.urlparse(host)
+        scheme = info.scheme if info.scheme else "https"
```

Comment:
Is this ever not https?  might simplify to make host be just the hostname, not the url, and always use https